File: 13-workflows/npm-package/workflow-control.ts
/**
 * Cost Katana Workflows: Pause, Resume, Cancel
 * 
 * Control workflow execution in real-time.
 * 
 * Run: npx ts-node 17-workflows/npm-package/workflow-control.ts
 */

import axios from 'axios';
import { executeWorkflow, getExecutionStatus } from './create-workflow';

const API_BASE = 'https://cost-katana-backend.store/api';
const API_KEY = process.env.COST_KATANA_API_KEY;

async function pauseWorkflow(executionId: string) {
  try {
    const response = await axios.post(
      `${API_BASE}/workflows/executions/${executionId}/pause`,
      {},
      {
        headers: { 'Authorization': `Bearer ${API_KEY}` }
      }
    );

    console.log('⏸️  Workflow paused');
    return response.data.data;
  } catch (error: any) {
    console.error('❌ Failed to pause:', error.response?.data || error.message);
    throw error;
  }
}

async function resumeWorkflow(executionId: string) {
  try {
    const response = await axios.post(
      `${API_BASE}/workflows/executions/${executionId}/resume`,
      {},
      {
        headers: { 'Authorization': `Bearer ${API_KEY}` }
      }
    );

    console.log('▶️  Workflow resumed');
    return response.data.data;
  } catch (error: any) {
    console.error('❌ Failed to resume:', error.response?.data || error.message);
    throw error;
  }
}

async function cancelWorkflow(executionId: string) {
  try {
    const response = await axios.post(
      `${API_BASE}/workflows/executions/${executionId}/cancel`,
      {},
      {
        headers: { 'Authorization': `Bearer ${API_KEY}` }
      }
    );

    console.log('⛔ Workflow cancelled');
    return response.data.data;
  } catch (error: any) {
    console.error('❌ Failed to cancel:', error.response?.data || error.message);
    throw error;
  }
}

async function workflowControlDemo() {
  console.log('🥷 Workflow Control Demo\n');

  // This would need a long-running workflow to demonstrate
  console.log('💡 Workflow Control Operations:');
  console.log('\n   Pause:');
  console.log('   - Pauses after current step completes');
  console.log('   - State is preserved');
  console.log('   - Can be resumed later');
  
  console.log('\n   Resume:');
  console.log('   - Continues from paused state');
  console.log('   - All context preserved');
  console.log('   - No data loss');
  
  console.log('\n   Cancel:');
  console.log('   - Stops execution immediately');
  console.log('   - Partial results saved');
  console.log('   - Cannot be resumed');

  console.log('\n📝 Example usage:');
  console.log(`
  // Start long-running workflow
  const execution = await executeWorkflow(templateId, variables);
  
  // Pause if needed
  await pauseWorkflow(execution.executionId);
  
  // Resume when ready
  await resumeWorkflow(execution.executionId);
  
  // Or cancel if no longer needed
  await cancelWorkflow(execution.executionId);
  `);
}

async function main() {
  if (!API_KEY) {
    console.error('❌ COST_KATANA_API_KEY required');
    process.exit(1);
  }

  await workflowControlDemo();
}

if (require.main === module) {
  main();
}

export { pauseWorkflow, resumeWorkflow, cancelWorkflow };

File: 13-workflows/python-sdk/workflow_control.py
"""
Cost Katana Workflows: Control Operations (Python)

Pause, resume, and cancel workflow executions.

Run: python 17-workflows/python-sdk/workflow_control.py
"""

import os
import requests

API_BASE = "https://api.costkatana.com/api"
API_KEY = os.getenv("COST_KATANA_API_KEY")

def pause_workflow(execution_id: str):
    """Pause a running workflow"""
    
    response = requests.post(
        f"{API_BASE}/agent-trace/executions/{execution_id}/pause",
        headers={"Authorization": f"Bearer {API_KEY}"}
    )
    
    response.raise_for_status()
    print("⏸️  Workflow paused")
    return response.json()["data"]

def resume_workflow(execution_id: str):
    """Resume a paused workflow"""
    
    response = requests.post(
        f"{API_BASE}/agent-trace/executions/{execution_id}/resume",
        headers={"Authorization": f"Bearer {API_KEY}"}
    )
    
    response.raise_for_status()
    print("▶️  Workflow resumed")
    return response.json()["data"]

def cancel_workflow(execution_id: str):
    """Cancel a workflow"""
    
    response = requests.post(
        f"{API_BASE}/agent-trace/executions/{execution_id}/cancel",
        headers={"Authorization": f"Bearer {API_KEY}"}
    )
    
    response.raise_for_status()
    print("⛔ Workflow cancelled")
    return response.json()["data"]

def main():
    if not API_KEY:
        print("❌ COST_KATANA_API_KEY required")
        return
    
    print("🥷 Workflow Control Operations (Python)\n")
    
    print("💡 Available operations:")
    print("\n   pause_workflow(execution_id)")
    print("   - Pauses after current step")
    print("   - State preserved")
    
    print("\n   resume_workflow(execution_id)")
    print("   - Continues from paused state")
    print("   - No data loss")
    
    print("\n   cancel_workflow(execution_id)")
    print("   - Stops immediately")
    print("   - Cannot resume")

if __name__ == "__main__":
    main()
